Basic imlementation of Scene

OskarG0 opened the pull request:
#10
Implemented a simple and basic structure for scenes. 
Did not manage to get clang format to work, so the code may differ.

joelsiks commented on src/engine/Scene.hpp:
```diff
@@ -22,12 +22,26 @@ class Scene : public renderer::IDrawable {
 
     std::multiset<std::shared_ptr<GameObject>, GameObjectComparator> objects;
 
+    int gridLenght;
+    int gridHeight;
+    bool dupsAllowed;
```
Prefix member variables with `m_`, i.e `m_gridLength`.

joelsiks commented on src/engine/Scene.hpp:
```diff
@@ -22,12 +22,26 @@ class Scene : public renderer::IDrawable {
+    void removeAllObject(std::shared_ptr<GameObject> object);
+    bool existObject(std::shared_ptr<GameObject> object);
+
+    int sizeObjects();
```
I feel this could be renamed to `numObjectsInScene` or something similar. Size means something different to me.

joelsiks commented on src/engine/Scene.hpp:
```diff
@@ -22,12 +22,26 @@ class Scene : public renderer::IDrawable {
+    int sizeObjects();
+    void setDupsAllowed(bool statement);
+
+    // bool collisionObject(std::shared_ptr<GameObject> object);
+    // std::vector<GameObject> detectObject(std::shared_ptr<GameObject>, int
+    // range);
```
remove?

joelsiks commented on src/engine/Scene.hpp:
```diff
@@ -22,12 +22,26 @@ class Scene : public renderer::IDrawable {
 
     std::multiset<std::shared_ptr<GameObject>, GameObjectComparator> objects;
```
I know Jakob added this, but perhaps we should discuss this in our next meeting.

joelsiks commented on src/test/scene.test.cpp:
```diff
@@ -0,0 +1,212 @@
+
+int main(int argc, char *argv[]) {
+
+    int res = test_add();
+    if (res) {
+        printf("test_add FAILURE\n");
+        return EXIT_FAILURE;
+    }
+    printf("test_add SUCCESS\n");
+
+    res = test_remove();
+    if (res) {
+        printf("test_remove FAILURE\n");
+        return EXIT_FAILURE;
+    }
+    printf("test_remove SUCCESS\n");
+
+    res = test_exist();
+    if (res) {
+        printf("test_exist FAILURE\n");
+        return EXIT_FAILURE;
+    }
+    printf("test_exist SUCCESS\n");
+
+    printf("all complete\n");
+    return EXIT_SUCCESS;
```
Instead of exiting on a failed test, which would mean not running subsequent tests, one could instead keep of how many tests have failed (if any) and print it in the end. If 0 failed test, all god, otherwise bad.

joelsiks reviewed (commented):
I have some comments on a few things but good stuff overall!

JoXW100 reviewed (changes requested):
Fond a severe issue, first look at the comments from @joelsiks. 

Why does the `Scene` object have `gridLenght` and `gridHeight` members? They seem to not be used for anything.

How does `multiset::sort` handle gameobjects with the same index? 
In my testing, this will treat the objects as being the same and fail the tests. This is not a good implementation as the index only decides the objects depth in the scene and is not unique.

Look into this issue.

OskarG0 commented:
Reworked the scene class to find object by the gameObjects pointer and their index.
The duplicate limit was also removed du to this change.
Unused varaibles was removed.
multiset do not have an explicit function set but uses it implicitly when inserting new values. The objects are then sorted in an order that is decided by the GameObjectComparator

JoXW100 reviewed (approved):
Looks good and all tests passed when testing.
